### ISSUE TITLE
Mintty wish list

### ISSUE THREAD

**Issue by mintty** [mintty/mintty#600]
Yet again I'm reserving a "round" issue number for a general mintty discussion topic.
The "Mintty discussion" issue (#500) wasn't too successful after a while but let's try once more.

**Comment by pmopmo** [comment]
I would like to have an option to tell mintty to strip tailing whitespace (especially line-breaks)  when pasting.

**Comment by mintty** [comment]
That's not transparent for all applications, some may expect pasted text verbatim. Also, stripping line breaks would smash words together, like:
is
this
desired?
-> "isthisdesired?"
I think such a function would be in the scope of applications.

**Comment by pmopmo** [comment]
Only the last white-space of an input should be stripped. So in your example only white-space after the question mark would be removed.

This request is mainly to stop accidentally running commands when copy/pasting

If I for example copied this below in mintty and wasn't paying attention I might select the whole line (that would include a newline)

`for N in 1 2 3; do echo $N; done`

And then when I paste it it gets executed. If trailing white-space was stripped only the command would be pasted and I would have to execute it myself by pressing newline.

I got used to this feature a few years ago when using iTerm2 on osx

**Comment by mintty** [comment]
So you only want to strip a final newline, not trailing space (which would not get copied in this case anyway). I think that's an option that can be introduced.

**Comment by kangjianbin** [comment]
Seperate UI code to make it easy to port to other platform (GTK/QT et, al).

**Comment by mintty** [comment]
Actually, X11 backend code was deliberately stripped off from putty when mintty was once derived as a dedicated easy-to-configure terminal for cygwin.
The platform-independent GUI framework is still largely intact, although later enhancements may break it sometimes because the abstraction was not verified anymore. Easy to repair, probably.
A little bit harder to handle, there were other enhancements about window integration, font configuration, character layout, and the like, that include Windows specifics more entwined with the terminal code.

However, I would be willing to support such a plan, especially if based on something portable already rather than X11 directly. But I'm not going to investigate that platform stuff myself. So my offer: if anyone familiar with, say, GTK, wants to craft a mintty backend for it, we can cooperate to port mintty to that backend.

**Comment by BrianInglis** [comment]
Suggest ignoring case in X colours (as recommended in X docs) by changing the `config.c(parse_colour)` comparison about line `532` to use `strcasecmp` instead of `strcmp`?
Suggest allowing American spelling `Color` as well as current Commonwealth English spelling `Colour` of config options in `config.c(options[])` around lines `163-283` by duplicating options with variant spelling, as they (and I) are unused to _colour_ being spelled properly in programs? ;^>
I encountered these issues when creating aliases to spawn mintty windows with reversed fg/bg colours from my defaults, for weather displays using ASCII art ( @chubin/wttr.in) and UTF-8  ( @ways/pyyrascii/graph.sh) symbols.

**Comment by mintty** [comment]
> ignoring case in X colours (as recommended in X docs)

Please point out the respective clause, then I'll adapt to it.

`Colour`: hesitating; this has been long tradition with mintty. I don't see what spelling has to do with weather reports; anyway, if you make scripts around tools, you'll always have to check the respective options precisely, nothing unusual here.

**Comment by BrianInglis** [comment]
On 2016-10-28 15:29, mintty wrote:
>     ignoring case in X colours (as recommended in X docs)
> Please point out the respective clause, then I'll adapt to it.

TL;DR: X colour names may be specified at least in: lower case
with or without spaces; and Title Case with or without spaces
e.g. antique white, antiquewhite, AntiqueWhite, Antique White
are all valid in most contexts; some may not allow for space
separated colour names. Most implementations ignore case.

My original O'Reilly X refs (with other historical reports
including Algol 60, 66, 68 and Pascal) were donated by my wife
or mother-law during basement cleanups, but the Internet Archive
has them online (below), and they state the colour names may be
used as given, or in mixed case with initial capitals, and spaces
removed.

In X11 R2 and R3 below, the names were given only in lower case
English, with both spellings of gray/grey, the intent being that
localized versions could be added or substituted; R4 added
CamelCaps spellings:

X11 Xlib Reference Manual Appendix D Color Database p537ff
	https://archive.org/stream/xlibretmanver1102nyemiss#page/536/mode/2up
X11 Xlib Reference Manual 2nd ed. Vol 2 Appendix D The Color Database p589ff
	https://archive.org/stream/xlibrefmanv115ed02nyemiss#page/588/mode/2up

The implementations however stored the names in dbm in lowercase,
and converted the requests to lowercase, and later the requirement
was added allowing the colour names to be given in ISO Latin1,
requiring the lowercase conversion to also handle Latin1 accented
characters, as can be seen in the logs for the following, which
were converted from original X.org RCS to git:

X11R6 oscolor.c
	https://cgit.freedesktop.org/~alanc/xc-historical/tree/xc/programs/Xserver/os/oscolor.c
X11R6 rgb.c
	https://cgit.freedesktop.org/~alanc/xc-historical/tree/xc/programs/rgb/rgb.c
X11R6 rgb.txt
	https://cgit.freedesktop.org/~alanc/xc-historical/tree/xc/programs/rgb/rgb.txt

Solaris also documents this.

X11 Solaris Color Name Database
	https://docs.oracle.com/cd/E19455-01/806-1363/serverintro-25/index.html

Wikipedia also mentions this but fails to cite it, although
demonstrates it by using the names in Title Case with spaces,
where those names are supported  by the colour names available,
as it states:

X11 Color Names
	https://en.wikipedia.org/wiki/X11_color_names

I still have a couple of X references, Johnson/Reichard 1992 using
only lower case in examples, with and without spaces, recommending
that spaces not be used when specifying X resources; the other uses
only Title Case with spaces in examples, as in Wikipedia.

Modern X server implementations build in the colour names and use
str[n]casecmp, as I suggested for mintty:

X server oscolor.c
	https://cgit.freedesktop.org/xorg/xserver/tree/os/oscolor.c

> |Colour|: hesitating; this has been long tradition with mintty. I
> don't see what spelling has to do with weather reports; anyway, if
> you make scripts around tools, you'll always have to check the
> respective options precisely, nothing unusual here.

As developers, we have become accustomed at the muscle memory level
to typing words differently from our common cultural usage, and as
our brains read words not letters, we may not pick up minor spelling
differences, unless it jars our cultural sensibilities as it
apparently does for some Americans and English.

X consistently uses the American spelling "color" so it would
accommodate those expecting X-like usage, like myself, and those
who learned English from American sources, unlike most Europeans,
similarly to supporting gray/grey in color names, which are both
accepted in American English and allowedin British English.
Consider it another addition to your interface I18N.

The weather reports were mentioned as the application reason for
spawning a mintty window, where both the resource name spelling
and the colour name capitalisation were unexpected interface
issues that could be ameliorated easily, with no downsides.

Unless you would prefer to automatically add the config ...Color
options from all existing ...Colour options as part of the build
process by moving the options into a separate file to allow e.g.
ADD_COLOR='s/^\(\s*{\s*"[^"]*\)Colour\(.*\)$/&\n\1Color\2/'
options.t:	options.txt
	sed "$$ADD_COLOR" $^ > $@
or something more elaborate to allow updating options in man pages,
configs, and other docs as part of the process.

This could be staged as three or more changesets: one to make
colour name comparison case insensitive, one to add Color config
options in config.c, another to split options into include file
say options.t, more to generate include file options.t from
options.txt, then also generate other related includes or files,
depending on how you want to architect and design the sources and
generated files.
You could provide guidelines and/or I could generate pull requests
and/or patches for changes and get your feedback.

-- 
Take care. Thanks, Brian Inglis, Calgary, Alberta, Canada

**Comment by mintty** [comment]
Colour name casing: actually my request was to
> Please point out the respective clause, then I'll adapt to it.

I've opened all your links but found nothing in an X spec except the Solaris one that states this. Anything in a common X spec would be convincing.

**Comment by BrianInglis** [comment]
On 2016-10-30 09:50, mintty wrote:
> Colour name casing: actually my request was to
>
>     Please point out the respective clause, then I'll adapt to it.
>
> I've opened all your links but found nothing in an X spec except the
> Solaris one that states this. Anything in a common X spec would be
> convincing.

The first two links to the Internet Archive Xlib refs for X11R2-4
Appendix D Color Names have the same statement in the last paragraph
on the right hand page, mentioned in the summary:

X11 Xlib Reference Manual Appendix D Color Database p537ff
	https://archive.org/stream/xlibretmanver1102nyemiss#page/536/mode/2up
X11 Xlib Reference Manual 2nd ed. Vol 2 Appendix D The Color Database p589ff
	https://archive.org/stream/xlibrefmanv115ed02nyemiss#page/588/mode/2up

The next three links demonstrate the original practice, and the last
link in the post the current practice, in the reference implementations
from X.org.

X11R6 oscolor.c
	https://cgit.freedesktop.org/~alanc/xc-historical/tree/xc/programs/Xserver/os/oscolor.c
X11R6 rgb.c
	https://cgit.freedesktop.org/~alanc/xc-historical/tree/xc/programs/rgb/rgb.c
X11R6 rgb.txt
	https://cgit.freedesktop.org/~alanc/xc-historical/tree/xc/programs/rgb/rgb.txt

X server oscolor.c
	https://cgit.freedesktop.org/xorg/xserver/tree/os/oscolor.c

-- 
Take care. Thanks, Brian Inglis, Calgary, Alberta, Canada

**Comment by mintty** [comment]
Ah, thanks. I hadn't read the text but tried to search for "case" which doesn't work in that presentation.
Two notes for the records:
* The wording "form shown or in mixed case" is actually already reflected in the database (e.g. both "blue violet" and "BlueViolet" are listed explicitly) but not completely (e.g. "Aquamarine" is not listed).
* Arbitrary or full upper-case (e.g. "BLUE") would strictly not be covered by the wording but that's a minor issue.

Will change as requested.

**Comment by BrianInglis** [comment]
On 2016-10-30 15:00, mintty wrote:
> Ah, thanks. I hadn't read the text but tried to search for "case"
> which doesn't work in that presentation.
> Two notes for the records:
>
> * The wording "form shown or in mixed case" is actually already
>   reflected in the database (e.g. both "blue violet" and "BlueViolet"
>   are listed explicitly) but not completely (e.g. "Aquamarine" is not
>   listed).
> * Arbitrary or full upper-case (e.g. "BLUE") would strictly not be
>   covered by the wording but that's a minor issue.
>
> Will change as requested.

Cheers!

Just managed to find the latest X.org docs, which have been reorganized
to move the DocBook sources to the same modules as the source code,
so colour name mapping is now documented with the functions:

https://www.x.org/releases/X11R7.7/doc/libX11/libX11/libX11.html#Mapping_Color_Names_to_Values

which all say "Use of uppercase or lowercase does not matter." following
the caveat "If the color name is not in the Host Portable Character Encoding,
the result is implementation-dependent." requiring ASCII names for consistency.

-- 
Take care. Thanks, Brian Inglis, Calgary, Alberta, Canada

**Comment by silverwind** [comment]
1) Tabs would be awesome to have.
2) I know it's been [denied](https://github.com/mintty/mintty/issues/468) once, but I still would like a option to disable the bold overstrike effect. I'm maintaining a fork just for that.

**Comment by mintty** [comment]
1. This is exhaustively discussed in #8 (duplicate). My last offer was that if someone provides a seamless integration with ConEmu, I would package that with the release.
2. Enforced bold mode was dropped already in 2.2.1 (#478). You just need to set option BoldColour.

**Comment by zwhfly** [comment]
I would like Mintty to have the ability to be minimized to a system tray icon, if that's possible.

**Comment by mintty** [comment]
The system tray is usually used by system tools which mintty isn't really, so I don't see the use case.
However, as a personal preference, I might consider to support that but unfortunately it's not another option of function `ShowWindow` so someone would have to find out how it works.

**Comment by zwhfly** [comment]
Actually, I just found another tool([RBTray](http://rbtray.sourceforge.net/)). It can minimize almost any window of other processes to the system tray. It works quite well for me.
It's GPL, and the source code is pretty slim(~400 lines). Roughly browsing of the code suggests that it uses [Shell_NotifyIcon](https://msdn.microsoft.com/en-us/library/windows/desktop/bb762159(v=vs.85).aspx).

**Comment by mintty** [comment]
Implementation might involve:
* capture the message indicating that the window is being minimized externally (e.g. by clicking the upper-right minimize button); likely WM_SYSCOMMAND (SC_MINIMIZE) in win_proc
* handle internal minimize action (win_set_iconic)
* activate the taskbar icon
* hide (and minimize?) the main window
* handle mouse click on the taskbar icon to restore the window
* equip the taskbar icon with a menu to handle multiple minimized windows

You are welcome to propose a patch; I would take care of the configuration.

**Comment by silverwind** [comment]
Maybe you can conside supporting [iTerm's image protocol](https://www.iterm2.com/documentation-images.html). It's a way for terminal to display images by parsing a special escape sequence. I's quite useful to quickly check out images on the shell.

**Comment by mintty** [comment]
That would now be possible with the img2sixel tool from https://github.com/saitoha/libsixel (unfortunately not a cygwin package yet, maybe you can ask @saitoha to build and submit a cygwin package...),
or with the netpbm package: convert image image.sixel; cat image.sixel; (there must be some way to tell netpbm convert to output directly to stdout but I don't see how).
For the iTerm protocol, one of those tooles (likely img2sixel) would have to be embedded into mintty, but maybe the tool solution is sufficient for you?

**Comment by silverwind** [comment]
I didn't know about Sixel yet, but isn't that format kind of limited, e.g. it cannot display 24 bit colors, but instead only a color-reduced image? Can mintty render raw 24-bit pixels yet?

**Comment by mintty** [comment]
I'm open to such a feature (the inline display option only) but I guess we'd need @saitoha to cooperate at least on the bitmap handling (storage, rendering and scrollback buffer).

**Comment by DArcMattr** [comment]
This is up there with the ligatures ticket, but also relates to the Sixel request, which is to display full color emoji in the terminal like how MacOS terminals can.

**Comment by mintty** [comment]
> Built-in tab support would be really great.

I guess you are aware of issue #8 and my previous comments both here and there.
Despite your valid use case, this is not among the core capabilities of a terminal.
Also it's a significant implementation endeavour.
Someone will have to implement a virtual window overlap approach (which I'd prefer).
And someone (likely not me) will have to implement the switching mechanism and GUI.

> • Multiple Mintty windows open (makes switching between terminals a hassle)

How about Ctrl+TAB / Ctrl+Shift+TAB?

**Comment by mintty** [comment]
@DArcMattr, are you aware that Sixel display has been supported for a year now?

**Comment by mintty** [comment]
@silverwind, @silentbugs, and all Tabs addicts:
I had some intuition how to conciliate the issue and implemented "Virtual Tabs". A proof of concept demo is available for download. A menu displays a switching list of all mintty windows.
For the demo, it's attached to the context menu and also reachable with Ctrl+Shift+I.
Other keys might be desirable, and probably it should not be in the default context menu, but maybe a separate one, to open with some modifier+mouse.

**Comment by silentbugs** [comment]
@mintty Nicely done, however this still controls the current "multi-window" behavior. I think a tab would imply that there is a single application window, and just some shortcuts to add, remove (?), and switch to previous/next tabs as well as context menus for those actions (e.g. Right click -> Add new tab).

Much like how you nested the current behavior inside the context menu, we could have a tab bar that's always visible, displaying the number of tabs. CTRL+Tab and CTRL+Shift+Tab could be bound to switching Tabs (rather than switching between windows) which is the default Windows behavior on a "tabbed" application. Example photo of [KDE's Konsole](http://images.shoutwiki.com/raju/3/30/Multiple_tabs_in_konsole.png) with tabs.

I don't know if and how that would be possible, but I'm sure it would be a really good addition.

The main issue of the current approach is that there's several windows of the same application, which in turn messes with Alt-Tab, Windows-Tab (presenting windows) and switching between the terminal and other applications. You can correct me if I'm wrong or if this would apply to a higher level application e.g. a wrapper around mintty (such as guake, yakuake and other drop-down terminals).

**Comment by mintty** [comment]
I assume the main use case is to easily switch between Tabs, or Virtual Tabs in this case. And that is supported via menu even better than with a tab bar because the window titles can be shown in full length, while a tab bar would often need to abbreviate. Ctrl+TAB would now behave exactly as you wish. A New session function is available (Alt+F2 or from the system menu), and I can of course put a copy next to the "Tabs" list.
I do not think, personally, that the feature of a single application window is still so much of an additional benefit, but it could be simulated to some extent by forcing switched or new window to the same location as the current one. However, what if you resize one of them? How do other terminals handle that, are all "tabs" implicitly resized, too?
If there are suggestions how to layout the tabs list in the various (modified) flavours of the context menu, I would appreciate; otherwise, I could perhaps make that configurable.

**Comment by silentbugs** [comment]
@mintty Yes, switching between (virtual) tabs easily is the current point, and at the moment is achieved by keyboard shortcuts and, with the change you made, via a context menu. I still believe that tabs should be visible and accessible to click to change, just for the sake of completeness of this feature. If possible the shortcuts to change to next/prev tab should be configurable as well.

Single terminal window, at least as an option, would be greatly appreciated by developers. It's really annoying to have a bunch of programs open, let alone another bunch of terminal windows. It makes switching between applications a lot messy.

On my Linux rig I've been using dropdown terminals for years, meaning that my terminal window is never a part of active windows but rather a separate application which I can bring to front whenever I need to with the hit of a key combination. Now, I obviously don't think that mintty should be a dropdown terminal as this is out of the application's scope as a terminal emulator.

As far as laying out the tab list, I don't believe it belongs in the context menu. It could be statically placed on the bottom of the window as a bar/list and we could then have an option to enable/disable said bar.

For the single application window, I guess if it's possible to add a "hack" to only display the current tab as a visible window it would do the trick, again toggled via option. When the "single terminal window" option is selected all windows should have one size, the current terminal's size. If we resize the current terminal, all terminals inherit that size. The behavior should be that of a single application window. That's how other terminal emulators handle this as well, since there's only one window. Resizing that window implicitly resizes all tabs.

I'm wondering if we minimize and bring to front the different windows if there is a way to disable the minimize/restore animation, since it would delay switching between tabs.

**Comment by mintty** [comment]
As I had suggested, some more approximation of a single-window view can be achieved. (True single-window implementation is out of reach with any reasonable design and implementation effort.)
The windows would be hidden rather than minimized, avoiding animation effects, or maybe simply overlayed.

However, I currently don't see how sessions started from one virtual single window could be distinguished from other mintty sessions, started separately. So they would be captured all into one (unless the Class parameter is explicitly used). That means, windows of different size and positions would then be merged into tabs of one virtual window. That raises some complication in the handling of position. A window that is being switched to could be repositioned to the current one.

Also, about resizing, what if some virtual tabs run at different font size (after zooming)?

About these open questions, some preferences or ideas would be appreciated.

**Comment by silentbugs** [comment]
@mintty my approach would be that since single-window view is going to be a workaround rather than an actual implementation, so everything should default to "the same as the main window" to prevent any issues, and then each item (e.g. different sizes, different zoom) can be investigated to see if it's possible to implement without causing any issues. Anything that is implemented could also be an option; for example: if you were to implement different position per window, you can add an option to choose if you'd like "all tabs in the same zoom" or "tab-specific zoom". I think this goes for all items (position, size, zoom..).

Different window sizes already caused trouble so all windows should inherit their size from the main virtual window.

> That means, windows of different size and positions would then be merged into tabs of one virtual window.

I believe that this is the desired behavior with tabs.

> That raises some complication in the handling of position. A window that is being switched to could be repositioned to the current one.

I believe all windows should inherit size and position from the main "virtual" window as you mentioned. Whenever that window is moved and/or resized, it affects all child windows.

> what if some virtual tabs run at different font size (after zooming)?

Do you think this could cause any problems? If so, default behavior becomes "all windows have the same font size" otherwise per-tab font size.

**Comment by mintty** [comment]
Released 2.8.0.

**Comment by mintty** [comment]
I've uploaded further levels of SessionGeomSync. Please check.
Not thoroughly tested yet: full-screen modes, multi-monitor behaviour.

**Comment by silentbugs** [comment]
@mintty Is there something in particular you'd like to test?

**Comment by mintty** [comment]
No, just the different SessionGeomSync levels with session switching, involving windows of different configuration (like maximized, full-screen, on differrent monitors).

**Comment by BrianInglis** [comment]
Please consider supporting XTerm compatible option -geometry WxH[[-+]X[-+]Y], as in Cygwin,  x-terminal-emulator is symlinked to mintty instead of XTerm, and use of -geometry is expected and understood, whereas -s WxH -p X,Y is unobvious. 
Option -geometry would be both: an alternative to -s WxH, and a combination of -s WxH -p X,Y; and allow X and Y to be negative, where -X is relative to left of screen and -Y is relative to bottom of screen. 
Using getopt_long_only instead of getopt_long allows use of single and/or double dashes before long option names where there is no ambiguity with short options: see e.g. https://linux.die.net/man/3/getopt_long_only or man 3 getopt_long_only for explanations.

**Comment by BrianInglis** [comment]
Please consider supporting end of line "# comments" in config files. 
Currently commented out default config lines have to be copied and split into separate comment and config variable definition lines, rather than just removing the leading "#" and changing the assigned value from the default. 
Please also consider continuing reading from the next config definition line when an error occurs, as it appears that reading from the config file stops when any definition error occurs.

**Comment by mintty** [comment]
I was considering a geometry option already, but that would likely be `--geometry` as a change to single-dash option could raise a bunch of compatibility problems.
End-of-line comments would be nice, I've just missed them myself.
An option error, however, does not skip the rest of the config file for me.

**Comment by BrianInglis** [comment]
Please consider generating -l, --log output with only Unix line terminators \n instead of the current DOS \r\n line terminators, in line with recent changes to other Cygwin Windows utilities to generate output with only Unix line terminators \n , after the base utilities changed to stop handling Windows line terminators transparently. Please also consider ensuring that any other multi-line output only uses \n terminators e.g. MINTTY_BUFFER mentioned in a recent issue.

**Comment by BrianInglis** [comment]
getopt_long_only() allows either one or two dashes to precede long options for compatibility of -geometry, --geometry.

**Comment by mintty** [comment]
I would agree to log terminal output verbatim (maybe even considering whether a CR was originally included). However, the CR seems to get inserted quite magically, while passing an LF from the child process to the terminal process through the pty. So it's quite tricky to do that change. Analysis welcome.
(Note that xterm also adds CR.)

**Comment by silentbugs** [comment]
@mintty about tabs: with the list of tabs that you added we're pretty much set, I'm going to list a few things as in descending order of importance (higher = more important): 

1) "single window" option, where other mintty windows are hidden
2) Previous & next tab keyboard shortcuts should be editable (currently ctrl+tab / ctrl+shift+tab)
3) Ideally the list of tabs that you implemented could also be added as a clickable menu, toggleable via option, possibly on the tab bar like web browser tabs: fixed width and once it runs out of space all tabs shrink to fit them in the screen.

**Comment by mintty** [comment]
> about tabs: with the list of tabs that you added we're pretty much set,

Thanks, I think I'll close issue 8.

> 1. "single window" option, where other mintty windows are hidden

With `SessionGeomSync=4`, other windows are virtually hidden already. What's missing?

> 2. Previous & next tab keyboard shortcuts should be editable (currently ctrl+tab / ctrl+shift+tab)

All shortcuts could be made editable, at the cost of adding complexity and potential confusion. What shortcuts would you have in mind specifically for tabs, to motivate such a change?

> 3. Ideally the list of tabs that you implemented could also be added as a clickable menu, toggleable via option, possibly on the tab bar like web browser tabs: fixed width and once it runs out of space all tabs shrink to fit them in the screen.

Yes, and perhaps as a dedicated process, in order to maintain the principle of the Virtual Tabs feature not to compromise the integrity of each terminal window.
As I had offered in https://github.com/mintty/mintty/issues/8#issuecomment-339660869, if someone contributes the GUI implementation, I will provide the functional backend, with communication and tab management as needed.

**Comment by silentbugs** [comment]
> With SessionGeomSync=4, other windows are virtually hidden already. What's missing?

The SessionGeometry options feels very hacky at the moment. The window only visually disappears, however it appears on both Task View and when hovering the mintty taskbar application. 

I think only the current window should be visible in all of the above scenarios. Also, with SessionGeometry=4, the position and size of the window should be the same, i.e. the current window's dimensions and size. I think this would bring this closer to a single window application.

> What shortcuts would you have in mind specifically for tabs, to motivate such a change?

Everyone has their own key bindings for switching tabs. I personally use yakuake's shift+left/right arrow.

**Comment by mintty** [comment]
> The window only visually disappears, however it appears on both Task View and when hovering the mintty taskbar application.

We could make it SessionGeomSync=5 to have those disappear... If someone provides an implementation to make it disappear; which is quite tricky.

In what scenario are position and size not the same for you?

**Comment by silentbugs** [comment]
> In what scenario are position and size not the same for you?

1. Open a terminal (default size).
2. Open a second terminal (default size).
3. Maximize  the second terminal.
4. Switch to first terminal (ctrl+tab).

The first terminal does not resize. Also it is visible in the background in case you have opacity.

Interestingly enough, both terminals seem like they are in "maximized" state because they both have the "Restore Down" button instead of maximize on the top bar. However, only the first window is full screen.

Also when switching tabs via the context menu, the tabs get resized to full width/height but not maximized.

----

I still don't think that this is the right approach to tab implementation, especially on Windows. There's an entire unutilized top bar that could very well be used for placing separate tabs on a single window, or even custom virtual tabs within the application. Working around multiple windows and trying to get them to work as tabs might be easy but imo is the wrong approach. I understand that proper tabs would probably be a lot of work and I wish my C skills were better in order to try and help.

**Comment by mintty** [comment]
The current Virtual Tabs implementation achieves what can be done with "normal" Windows means (some gaps like with full-screen may still be checked). Let me summarize again two major obstacles to getting more:
* Extremely high effort to fiddle out any "tricky" ways of using Windows to achieve special effects (like hiding taskbar icons). I've been torturing myself with insane Windows APIs for a number of other, more essential features already, and I'm not going to repeat that for an issue that I personally do not consider essential.
* Integrity of terminal operation. Sure there would be the option of running only one Windows window in the first place, but that would need a lot of intertwining the tabs code with the existing mintty code. I am not willing to take the risk to raise a number of stability issues and take according complaints.

I hope these constraints are understood and accepted.

**Comment by mkq** [comment]
I'm late to this tabs discussion because I was just watching #8, not this issue. Anyway, my 2 cents:

> The window only visually disappears, however it appears on both Task View and when hovering the mintty taskbar application.

I don't know if possible at all, but a single Alt+Tab entry, but separate taskbar buttons would actually be my favorite mix. The latter is handy to keep using `BellTaskbar=yes`, to get notified of a session being back at the prompt for example.

**Comment by mintty** [comment]
You are invited to explore whether this is possible with the Windows API :)

**Comment by mintty** [comment]
@BrianInglis: We could support single-dash long options conditionally (in order to avoid clashes with multiple short options), after setting an option in the config file. This could be called, for example, `ShortLongOpts` or `SingleDashLongOptions`; comments or suggestions?

**Comment by mintty** [comment]
Released 2.8.2 with some recently requested enhancements.

**Comment by mintty** [comment]
@DArcMattr: The current repository upload provides full emoji support.

**Comment by mintty** [comment]
Released 2.8.4 with full emoji support.

**Comment by JonathanHolvey** [comment]
Could a single window mode be a possible future feature? This would add the tidiness of having all the 'virtual tabs' in the same place, without the added GUI complexity of actual tabs. A simple `[1/3]` at the beginning of the title bar would go a long way to help keep track of multiple consoles when <kbd>Alt</kbd>+<kbd>Tab</kbd>ing between them.

**Comment by mintty** [comment]
You are aware that with `SessionGeomSync=4` the Virtual Tabs feature accomplishes "virtually" what you are asking for? If it doesn't for a specific situation, please describe it and I'll look into whether it can be improved. I'm not sure what you mean with "GUI complexity" - enabling Virtual Tabs is basically simple (but not available from the Options menu).
In general, this feature is about the best that can be achieved under the conditions I mentioned elsewhere, just a "Tabs Bar" is missing; the framework for it (maintaining the tabs information) is already available...

**Comment by silentbugs** [comment]
@mintty @JonathanHolvey current implementation (`SessionGeomSync=4`) emulates single window functionality without actually achieving it. The taskbar still shows multiple windows open, each window's location and size are not maintained. What both Jonathan and myself are looking for is a simple tabbed process, much like browser tabs: a single window (location on screen, maximized, minimized etc.) with multiple tabs within that window.

A few weekends ago I created a simple windows tabbed application and tried to launch mintty within each tab, but I couldn't change each tab's content nicely without hacking around. That was a single day attempt at an API that I'm completely unfamiliar with, I'm sure someone with experience in that could achieve this in a couple of hours.

**Comment by mintty** [comment]
> each window's location and size are not maintained

See my previous comment:  If it doesn't for a specific situation, please describe it and I'll look into whether it can be improved.

> What both Jonathan and myself are looking for ...

I know what you're looking for but consider the implications, see e.g. https://github.com/mintty/mintty/issues/8#issuecomment-194838312

> I couldn't change each tab's content nicely without hacking around.

That's the point. The issue is not just about creating multiple tabs but also maintaining each tab's integrity as a terminal.

> I'm sure someone with experience in that could achieve this in a couple of hours.

Could be. So please hire someone with that experience. And experience with terminal code at the same time, for that matter.

**Comment by JonathanHolvey** [comment]
@mintty I wasn't aware of the `SessionGeomSync=4` option. You're right, it does *virtually* what I was asking for, albeit in a fairly clunky way. All the windows are moved to the same location after a mouse event on the window border (I assume).

It does seem to have problems with keyboard shortcuts for multi-monitor setups; for example <kbd>Win</kbd>+<kbd>Shift</kbd>+<kbd>&rarr;</kbd> (on Windows 7 at least) will split the location of two windows until one is focused using the mouse.

Personally, I think you're right not to clutter the UI with actual tabs (this is what I meant by GUI complexity), since 99% of input is achieved using the keyboard when using a console program. However, it would be nice to have all the 'tabs' living in a single window.

I understand there are restrictions that make this difficult to achieve, and that I'm not pressuring for it to be implemented. I just wanted to get it down onto the wish list, and to find out if it would be easier than an actual tabbed interface.

Cheers :smiley:

**Comment by JonathanHolvey** [comment]
Incidentally, I've just discovered that it is actually possible to run Mintty *inside* ConEmu, as a [ChildGUI](https://conemu.github.io/en/ChildGui.html). This means that ConEmu can be used to manage multiple Mintty windows as tabs, provided they are spawned using a command external to ConEmu and Mintty, or using ConEmu's context menus.

Most of ConEmu's keybindings are overridden and passed directly to the child process (Mintty in this case) so <kbd>Ctrl</kbd>+<kbd>T</kbd> doesn't work to create a new tab. <kbd>Alt</kbd>+<kbd>F2</kbd> results in a new Mintty window outside of ConEmu, so that can't be used either.

**Comment by mkq** [comment]
I also gave ConEmu a (second) try a few months ago and am quite happy with it since. My workaround for tab switching is:

- mintty: disable Ctrl+[Shift+]Tab
- ConEmu: enable Main / Tab bar / Internal CtrlTab
- ConEmu: under Keys & Macro, bind "Set focus to ConEmu" to Alt+Shift+Backspace
and "Set focus to child GUI application" to Ctrl+Shift+Backspace
- run the following Autohotkey script; I included three sets of hotkeys: [Ctrl+Tab] / [Ctrl + Shift + Tab], [Ctrl + PageDown] / [Ctrl + PageUp], [Ctrl + Period] / [Ctrl + Comma]. The latter are the ones I actually use; the others are barely tested but should work.

```
#SingleInstance force

#IfWinActive ahk_class VirtualConsoleClass ahk_exe ConEmu64.exe
^+Tab::
^PgUp::
^,::
	BlockInput Send
	Send !+{backspace}
	Sleep 4
	Send ^+{tab}^+{backspace}
	BlockInput Default
	return
^Tab::
^PgDn::
^.::
	BlockInput Send
	Send !+{backspace}
	Sleep 4
	Send ^{tab}^+{backspace}
	BlockInput Default
	return

#IfWinActive
```

I don't open new tabs often, so I don't mind using the mouse for it. But you could do [Alt+Shift+Backspace], [Alt + Space] to open ConEmu's system menu, and then hit [Down], [Enter].

**Comment by mintty** [comment]
I was struck back more than once already by ConEmu's unwieldy configuration. Also it has other drawbacks; mouse message handling is buggy (see #724), Alt+F2 does not behave well.
If you have a working ConEmu configuration that's fine for you. I don't think it's in a state to be promoted in the mintty project, though.

**Comment by mintty** [comment]
Released 2.8.5 which supports emoji style precedence using text attribute SGR 51/52 and displays emoji sequence short names in Character Info mode.

**Comment by mintty** [comment]
@JonathanHolvey, I've tried to enable Win+Shift moving of tab sets (if SessionGeomSync ≥ 2); uploaded to repository. Please try if possible.

**Comment by mintty** [comment]
I'm closing this issue which has meanwhile become too long to keep an overview, in favour of a new general chat issue #777.
Feel free to enter any remaining questions or proposals again there or in a new issue.

**Comment by mintty** [comment]
Released 2.9.0.

**Comment by silentbugs** [comment]
@mintty I'd like to let you all know that windows will be adding a feature called "Sets" which will allow for tabbing multiple applications and it's exactly what we were looking for with mintty's tabs. Here's an [early preview video](https://www.youtube.com/watch?v=3lEjuU-XFHg) by Microsoft demonstrating Sets.

**Comment by Biswa96** [comment]
Sets was removed from Windows Insiders Build 17704. See this https://blogs.windows.com/windowsexperience/2018/06/27/announcing-windows-10-insider-preview-build-17704/

**Comment by silentbugs** [comment]
@Biswa96 It was only removed from the current release for further development, it wasn't cancelled, which means we can ignore tabs as Microsoft will be kind enough to provide it for us.
